pageserver/src/bin/pageserver.rs

```diff
@@ -150,70 +146,57 @@ fn main() -> anyhow::Result<()> {
 }
 
 fn initialize_config(
+    identity_file_path: &Utf8Path,
     cfg_file_path: &Utf8Path,
-    arg_matches: clap::ArgMatches,
     workdir: &Utf8Path,
-) -> anyhow::Result<ControlFlow<(), &'static PageServerConf>> {
-    let init = arg_matches.get_flag("init");
-
-    let file_contents: Option<toml_edit::Document> = match std::fs::File::open(cfg_file_path) {
+) -> anyhow::Result<&'static PageServerConf> {
+    // The deployment orchestrator writes out an indentity file containing the node id
+    // for all pageservers. This file is the source of truth for the node id. In order
+    // to allow for rolling back pageserver releases, the node id is also included in
+    // the pageserver config that the deployment orchestrator writes to disk for the pageserver.
+    // A rolled back version of the pageserver will get the node id from the pageserver.toml
+    // config file.
+    let identity = match std::fs::File::open(identity_file_path) {
         Ok(mut f) => {
-            if init {
-                anyhow::bail!("config file already exists: {cfg_file_path}");
+            let md = f.metadata().context("stat config file")?;
+            if !md.is_file() {
+                anyhow::bail!("Pageserver found identity file but it is a dir entry: {identity_file_path}. Aborting start up ...");
             }
+
+            let mut s = String::new();
+            f.read_to_string(&mut s).context("read identity file")?;
+            let doc: toml_edit::Document = s.parse().context("parse identity file toml")?;
+
+            PageserverIdentity::parse_and_validate(&doc)?
```

Comment:
Why not make `PageserverIdentity` derive serde deserialize & with `deny_unknown_fields`  ?

Reply:
No good reason. Was following the `PageserverConf` temaplate. Don't feel strongly either way. Can change it if you want.

Reply:
Would prefer it changed, I'm moving us further away from the handcrafted toml deserializer in https://github.com/neondatabase/neon/pull/7656

Reply:
[4087a1f](https://github.com/neondatabase/neon/pull/7766/commits/4087a1fdfe6f16346e70c60af2e2c8af1c08a075)